File: backend/src/routes/gemini.ts
import { Router, Request, Response } from 'express';
import { getSuspectResponse, generateCaseSummary, getOfficerChatResponse, getPartnerIntervention, getBadCopHint } from '../services/geminiChat.js';
import { generateCaseFromPrompt, checkCaseConsistency, editCaseWithPrompt, calculateDifficulty, applyConsistencyImagePipeline } from '../services/geminiCase.js';
import { generateImageRaw, generateEvidenceImage, generateEmotionalVariantsFromBase, generateOnePortraitVariantFromBase, generateSuspectFromUpload, generateNeutralPortraitForSuspect, regenerateSingleSuspect, pregenerateCaseImages, createImageFromPrompt, editImageWithPrompt } from '../services/geminiImages.js';
import { generateTTS } from '../services/geminiTTS.js';

const router = Router();

// --- CHAT ENDPOINTS ---

router.post('/chat/suspect', async (req: Request, res: Response) => {
  try {
    const { suspect, caseData, userInput, type, evidenceAttachment, currentAggravation, isFirstTurn, discoveredEvidence, currentGameTime, conversationHistory } = req.body;
    const result = await getSuspectResponse(suspect, caseData, userInput, type, evidenceAttachment, currentAggravation, isFirstTurn, discoveredEvidence, currentGameTime, conversationHistory);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] chat/suspect error:', error);
    res.status(500).json({ error: error.message || 'Failed to get suspect response' });
  }
});

router.post('/chat/officer', async (req: Request, res: Response) => {
  try {
    const { caseData, userMessage, evidenceFound, notes, chatHistory, timelineKnown, officerThread } = req.body;
    const result = await getOfficerChatResponse(
      caseData,
      userMessage,
      evidenceFound,
      notes,
      chatHistory,
      Array.isArray(timelineKnown) ? timelineKnown : [],
      Array.isArray(officerThread) ? officerThread : []
    );
    res.json({ text: result });
  } catch (error: any) {
    console.error('[Gemini Route] chat/officer error:', error);
    res.status(500).json({ error: error.message || 'Failed to get officer response' });
  }
});

router.post('/chat/partner', async (req: Request, res: Response) => {
  try {
    const { type, suspect, caseData, history, discoveredEvidence, timelineKnown } = req.body;
    const result = await getPartnerIntervention(
      type,
      suspect,
      caseData,
      history,
      discoveredEvidence,
      Array.isArray(timelineKnown) ? timelineKnown : []
    );
    res.json({ text: result });
  } catch (error: any) {
    console.error('[Gemini Route] chat/partner error:', error);
    res.status(500).json({ error: error.message || 'Failed to get partner intervention' });
  }
});

router.post('/chat/badcop-hint', async (req: Request, res: Response) => {
  try {
    const { suspect, discoveredEvidence, responseText } = req.body;
    const result = await getBadCopHint(suspect, Array.isArray(discoveredEvidence) ? discoveredEvidence : [], responseText);
    res.json({ text: result });
  } catch (error: any) {
    console.error('[Gemini Route] chat/badcop-hint error:', error);
    res.status(500).json({ error: error.message || 'Failed to get bad cop hint' });
  }
});

router.post('/chat/case-summary', async (req: Request, res: Response) => {
  try {
    const { caseData, accusedId, gameResult, evidenceDiscovered } = req.body;
    const result = await generateCaseSummary(caseData, accusedId, gameResult, evidenceDiscovered);
    res.json({ text: result });
  } catch (error: any) {
    console.error('[Gemini Route] chat/case-summary error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate case summary' });
  }
});

// --- CASE ENDPOINTS ---

router.post('/case/generate', async (req: Request, res: Response) => {
  try {
    const { userPrompt, isLucky } = req.body;
    const result = await generateCaseFromPrompt(userPrompt, isLucky);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] case/generate error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate case' });
  }
});

router.post('/case/consistency', async (req: Request, res: Response) => {
  try {
    const { caseData, baseline, editContext } = req.body;
    const result = await checkCaseConsistency(caseData, undefined, baseline, editContext);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] case/consistency error:', error);
    res.status(500).json({ error: error.message || 'Failed to check case consistency' });
  }
});

/** Narrative-only phase (no image regeneration). Pair with POST /case/consistency/images. */
router.post('/case/consistency/narrative', async (req: Request, res: Response) => {
  try {
    const { caseData, baseline, editContext } = req.body;
    const result = await checkCaseConsistency(caseData, undefined, baseline, editContext, { narrativeOnly: true });
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] case/consistency/narrative error:', error);
    res.status(500).json({ error: error.message || 'Failed narrative consistency check' });
  }
});

/** Image pipeline after narrative merge. Body: { mergedCase, originalCaseData } (original = pre-check draft). */
router.post('/case/consistency/images', async (req: Request, res: Response) => {
  try {
    const { mergedCase, originalCaseData } = req.body;
    if (!mergedCase || !originalCaseData) {
      res.status(400).json({ error: 'mergedCase and originalCaseData are required' });
      return;
    }
    const { changesMade } = await applyConsistencyImagePipeline(mergedCase, originalCaseData, undefined);
    res.json({ updatedCase: mergedCase, imagePipelineChanges: changesMade });
  } catch (error: any) {
    console.error('[Gemini Route] case/consistency/images error:', error);
    res.status(500).json({ error: error.message || 'Failed consistency image pipeline' });
  }
});

router.post('/case/edit', async (req: Request, res: Response) => {
  try {
    const { caseData, userPrompt, baseline } = req.body;
    const result = await editCaseWithPrompt(caseData, userPrompt, undefined, baseline);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] case/edit error:', error);
    res.status(500).json({ error: error.message || 'Failed to edit case' });
  }
});

// --- IMAGE ENDPOINTS ---

router.post('/image/generate', async (req: Request, res: Response) => {
  try {
    const { prompt, aspectRatio, refImages, mode, modelOverride } = req.body;
    const result = await generateImageRaw(prompt, aspectRatio, refImages, mode, modelOverride);
    res.json({ base64: result });
  } catch (error: any) {
    console.error('[Gemini Route] image/generate error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate image' });
  }
});

router.post('/image/evidence', async (req: Request, res: Response) => {
  try {
    const { evidence, caseId, userId, refImage, forDeceasedVictim, caseTheme } = req.body;
    const result = await generateEvidenceImage(evidence, caseId, userId, refImage, {
      forDeceasedVictim: !!forDeceasedVictim,
      caseTheme: typeof caseTheme === 'string' ? caseTheme : undefined,
    });
    res.json({ url: result });
  } catch (error: any) {
    console.error('[Gemini Route] image/evidence error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate evidence image' });
  }
});

router.post('/image/variants', async (req: Request, res: Response) => {
  try {
    const { neutralBase64, suspect, caseId, userId, caseTheme } = req.body;
    const result = await generateEmotionalVariantsFromBase(neutralBase64, suspect, caseId, userId, {
      caseTheme: typeof caseTheme === 'string' ? caseTheme : undefined,
    });
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] image/variants error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate emotional variants' });
  }
});

router.post('/image/variant-one', async (req: Request, res: Response) => {
  try {
    const { neutralBase64, variantKey, suspect, caseId, userId, caseTheme } = req.body;
    if (!variantKey || typeof variantKey !== 'string') {
      res.status(400).json({ error: 'variantKey is required' });
      return;
    }
    const result = await generateOnePortraitVariantFromBase(neutralBase64, variantKey, suspect, caseId, userId, {
      caseTheme: typeof caseTheme === 'string' ? caseTheme : undefined,
    });
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] image/variant-one error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate portrait variant' });
  }
});

router.post('/image/suspect-upload', async (req: Request, res: Response) => {
  try {
    const { suspect, userImageBase64, caseId, userId } = req.body;
    const result = await generateSuspectFromUpload(suspect, userImageBase64, caseId, userId);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] image/suspect-upload error:', error);
    res.status(500).json({ error: error.message || 'Failed to process suspect upload' });
  }
});

router.post('/image/regenerate', async (req: Request, res: Response) => {
  try {
    const { suspect, caseId, userId, theme } = req.body;
    const result = await regenerateSingleSuspect(suspect, caseId, userId, theme);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] image/regenerate error:', error);
    res.status(500).json({ error: error.message || 'Failed to regenerate suspect' });
  }
});

router.post('/image/regenerate-neutral', async (req: Request, res: Response) => {
  try {
    const { suspect, caseId, userId, theme } = req.body;
    const result = await generateNeutralPortraitForSuspect(suspect, caseId, userId, theme);
    res.json(result);
  } catch (error: any) {
    console.error('[Gemini Route] image/regenerate-neutral error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate neutral portrait' });
  }
});

router.post('/image/pregenerate', async (req: Request, res: Response) => {
  try {
    const { caseData, userId } = req.body;
    await pregenerateCaseImages(caseData, userId);
    // Return the mutated caseData (images are set directly on the object)
    res.json(caseData);
  } catch (error: any) {
    console.error('[Gemini Route] image/pregenerate error:', error);
    res.status(500).json({ error: error.message || 'Failed to pregenerate case images' });
  }
});

router.post('/image/create', async (req: Request, res: Response) => {
  try {
    const { userPrompt, aspectRatio } = req.body;
    const result = await createImageFromPrompt(userPrompt, aspectRatio);
    res.json({ base64: result });
  } catch (error: any) {
    console.error('[Gemini Route] image/create error:', error);
    res.status(500).json({ error: error.message || 'Failed to create image' });
  }
});

router.post('/image/edit', async (req: Request, res: Response) => {
  try {
    const { baseImageBase64, userPrompt, aspectRatio } = req.body;
    const result = await editImageWithPrompt(baseImageBase64, userPrompt, aspectRatio);
    res.json({ base64: result });
  } catch (error: any) {
    console.error('[Gemini Route] image/edit error:', error);
    res.status(500).json({ error: error.message || 'Failed to edit image' });
  }
});

// --- TTS ENDPOINT ---

router.post('/tts', async (req: Request, res: Response) => {
  try {
    const { text, voiceName, stylePrompt } = req.body;
    const base64Audio = await generateTTS(text, voiceName, stylePrompt);
    res.json({ audio: base64Audio });
  } catch (error: any) {
    console.error('[Gemini Route] tts error:', error);
    res.status(500).json({ error: error.message || 'Failed to generate TTS' });
  }
});

export default router;

File: backend/app/routes/gemini.py
"""Gemini AI routes — /api/gemini/*
Faithful port of src/routes/gemini.ts.
"""
from __future__ import annotations
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..services.gemini_chat import (
    get_suspect_response,
    get_case_summary,
    get_officer_response,
    get_partner_intervention,
    get_bad_cop_hint,
)
from ..services.gemini_case_core import (
    check_case_consistency,
    edit_case_with_prompt,
    generate_case_from_prompt,
    apply_consistency_image_pipeline,
)
from ..services.gemini_images import (
    generate_image_raw,
    generate_evidence_image,
    generate_emotional_variants_from_base,
    generate_one_portrait_variant_from_base,
    generate_suspect_from_upload,
    generate_neutral_portrait_for_suspect,
    regenerate_single_suspect,
    pregenerate_case_images,
    create_image_from_prompt,
    edit_image_with_prompt,
)
from ..services.gemini_tts import generate_tts

router = APIRouter(prefix="/api/gemini")


# --- CHAT ENDPOINTS ---

@router.post("/chat/suspect")
async def chat_suspect(request: Request):
    body = await request.json()
    try:
        # Build game_state from the flat fields the frontend sends
        game_state = {
            "aggravation": body.get("currentAggravation", 0),
            "isFirstTurn": body.get("isFirstTurn", False),
            "revealedEvidence": [e.get("title", "") for e in (body.get("discoveredEvidence") or []) if isinstance(e, dict)],
            "currentGameTime": body.get("currentGameTime"),
            "evidenceAttachment": body.get("evidenceAttachment"),
            "type": body.get("type", "talk"),
        }
        chat_history = body.get("conversationHistory") or []
        result = await get_suspect_response(
            body.get("suspect"), body.get("caseData"),
            game_state, chat_history, body.get("userInput", ""),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] chat/suspect error: {e}")
        return JSONResponse({"error": str(e) or "Failed to get suspect response"}, status_code=500)


@router.post("/chat/officer")
async def chat_officer(request: Request):
    body = await request.json()
    try:
        timeline_known = body.get("timelineKnown", [])
        if not isinstance(timeline_known, list):
            timeline_known = []
        officer_thread = body.get("officerThread", [])
        if not isinstance(officer_thread, list):
            officer_thread = []
        # Build game_state from the flat fields
        game_state = {
            "evidenceFound": body.get("evidenceFound", []),
            "notes": body.get("notes", {}),
            "chatHistory": body.get("chatHistory", {}),
            "timelineKnown": timeline_known,
            "officerThread": officer_thread,
        }
        result = await get_officer_response(
            body.get("caseData"), game_state,
            officer_thread, body.get("userMessage", ""),
        )
        return {"text": result}
    except Exception as e:
        print(f"[Gemini Route] chat/officer error: {e}")
        return JSONResponse({"error": str(e) or "Failed to get officer response"}, status_code=500)


@router.post("/chat/partner")
async def chat_partner(request: Request):
    body = await request.json()
    try:
        timeline_known = body.get("timelineKnown", [])
        if not isinstance(timeline_known, list):
            timeline_known = []
        # Build game_state from the flat fields
        game_state = {
            "aggravation": body.get("currentAggravation", 0),
            "discoveredEvidence": body.get("discoveredEvidence", []),
            "timelineKnown": timeline_known,
        }
        result = await get_partner_intervention(
            body.get("caseData"), body.get("suspect"),
            game_state, body.get("history", []),
            body.get("type", "goodCop"),
        )
        return {"text": result}
    except Exception as e:
        print(f"[Gemini Route] chat/partner error: {e}")
        return JSONResponse({"error": str(e) or "Failed to get partner intervention"}, status_code=500)


@router.post("/chat/badcop-hint")
async def chat_badcop_hint(request: Request):
    body = await request.json()
    try:
        discovered = body.get("discoveredEvidence", [])
        if not isinstance(discovered, list):
            discovered = []
        # Build game_state and chat_history from the flat fields
        game_state = {
            "discoveredEvidence": discovered,
        }
        result = await get_bad_cop_hint(
            body.get("caseData", {}), body.get("suspect"),
            game_state, body.get("chatHistory", []),
        )
        return {"text": result}
    except Exception as e:
        print(f"[Gemini Route] chat/badcop-hint error: {e}")
        return JSONResponse({"error": str(e) or "Failed to get bad cop hint"}, status_code=500)


@router.post("/chat/case-summary")
async def chat_case_summary(request: Request):
    body = await request.json()
    try:
        # Build game_state and accusation from the flat fields
        game_state = {
            "evidenceFound": len(body.get("evidenceDiscovered", [])),
            "suspectsSpoken": 0,
            "timelineFound": 0,
        }
        accusation = {
            "accusedIds": [body.get("accusedId")] if body.get("accusedId") else [],
        }
        result = await get_case_summary(
            body.get("caseData"), game_state, accusation,
        )
        return {"text": result}
    except Exception as e:
        print(f"[Gemini Route] chat/case-summary error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate case summary"}, status_code=500)


# --- CASE ENDPOINTS ---

@router.post("/case/generate")
async def case_generate(request: Request):
    body = await request.json()
    try:
        result = await generate_case_from_prompt(body.get("userPrompt", ""), body.get("isLucky", False))
        return result
    except Exception as e:
        print(f"[Gemini Route] case/generate error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate case"}, status_code=500)


@router.post("/case/consistency")
async def case_consistency(request: Request):
    body = await request.json()
    try:
        result = await check_case_consistency(
            body.get("caseData"), None, body.get("baseline"), body.get("editContext"),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] case/consistency error: {e}")
        return JSONResponse({"error": str(e) or "Failed to check case consistency"}, status_code=500)


@router.post("/case/consistency/narrative")
async def case_consistency_narrative(request: Request):
    """Narrative-only phase (no image regeneration). Pair with POST /case/consistency/images."""
    body = await request.json()
    try:
        result = await check_case_consistency(
            body.get("caseData"), None, body.get("baseline"),
            body.get("editContext"), {"narrativeOnly": True},
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] case/consistency/narrative error: {e}")
        return JSONResponse({"error": str(e) or "Failed narrative consistency check"}, status_code=500)


@router.post("/case/consistency/images")
async def case_consistency_images(request: Request):
    """Image pipeline after narrative merge. Body: { mergedCase, originalCaseData }."""
    body = await request.json()
    merged_case = body.get("mergedCase")
    original_case_data = body.get("originalCaseData")
    if not merged_case or not original_case_data:
        return JSONResponse({"error": "mergedCase and originalCaseData are required"}, status_code=400)
    try:
        result = await apply_consistency_image_pipeline(merged_case, original_case_data)
        return {"updatedCase": merged_case, "imagePipelineChanges": result.get("changesMade", [])}
    except Exception as e:
        print(f"[Gemini Route] case/consistency/images error: {e}")
        return JSONResponse({"error": str(e) or "Failed consistency image pipeline"}, status_code=500)


@router.post("/case/edit")
async def case_edit(request: Request):
    body = await request.json()
    try:
        result = await edit_case_with_prompt(
            body.get("caseData"), body.get("userPrompt", ""),
            None, body.get("baseline"),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] case/edit error: {e}")
        return JSONResponse({"error": str(e) or "Failed to edit case"}, status_code=500)


# --- IMAGE ENDPOINTS ---

@router.post("/image/generate")
async def image_generate(request: Request):
    body = await request.json()
    try:
        result = await generate_image_raw(
            body.get("prompt", ""), body.get("aspectRatio"),
            body.get("refImages"), body.get("mode"), body.get("modelOverride"),
        )
        return {"base64": result}
    except Exception as e:
        print(f"[Gemini Route] image/generate error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate image"}, status_code=500)


@router.post("/image/evidence")
async def image_evidence(request: Request):
    body = await request.json()
    try:
        result = await generate_evidence_image(
            body.get("evidence"), body.get("caseId"), body.get("userId"),
            body.get("refImage"),
            {
                "forDeceasedVictim": bool(body.get("forDeceasedVictim")),
                "caseTheme": body.get("caseTheme") if isinstance(body.get("caseTheme"), str) else None,
            },
        )
        return {"url": result}
    except Exception as e:
        print(f"[Gemini Route] image/evidence error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate evidence image"}, status_code=500)


@router.post("/image/variants")
async def image_variants(request: Request):
    body = await request.json()
    try:
        result = await generate_emotional_variants_from_base(
            body.get("neutralBase64", ""), body.get("suspect"),
            body.get("caseId", ""), body.get("userId", ""),
            {"caseTheme": body.get("caseTheme") if isinstance(body.get("caseTheme"), str) else None},
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] image/variants error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate emotional variants"}, status_code=500)


@router.post("/image/variant-one")
async def image_variant_one(request: Request):
    body = await request.json()
    variant_key = body.get("variantKey")
    if not variant_key or not isinstance(variant_key, str):
        return JSONResponse({"error": "variantKey is required"}, status_code=400)
    try:
        result = await generate_one_portrait_variant_from_base(
            body.get("neutralBase64", ""), variant_key, body.get("suspect"),
            body.get("caseId", ""), body.get("userId", ""),
            {"caseTheme": body.get("caseTheme") if isinstance(body.get("caseTheme"), str) else None},
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] image/variant-one error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate portrait variant"}, status_code=500)


@router.post("/image/suspect-upload")
async def image_suspect_upload(request: Request):
    body = await request.json()
    try:
        result = await generate_suspect_from_upload(
            body.get("suspect"), body.get("userImageBase64"),
            body.get("caseId"), body.get("userId"),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] image/suspect-upload error: {e}")
        return JSONResponse({"error": str(e) or "Failed to process suspect upload"}, status_code=500)


@router.post("/image/regenerate")
async def image_regenerate(request: Request):
    body = await request.json()
    try:
        result = await regenerate_single_suspect(
            body.get("suspect"), body.get("caseId"),
            body.get("userId"), body.get("theme", "Noir"),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] image/regenerate error: {e}")
        return JSONResponse({"error": str(e) or "Failed to regenerate suspect"}, status_code=500)


@router.post("/image/regenerate-neutral")
async def image_regenerate_neutral(request: Request):
    body = await request.json()
    try:
        result = await generate_neutral_portrait_for_suspect(
            body.get("suspect"), body.get("caseId"),
            body.get("userId"), body.get("theme", "Noir"),
        )
        return result
    except Exception as e:
        print(f"[Gemini Route] image/regenerate-neutral error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate neutral portrait"}, status_code=500)


@router.post("/image/pregenerate")
async def image_pregenerate(request: Request):
    body = await request.json()
    case_data = body.get("caseData")
    user_id = body.get("userId")
    try:
        await pregenerate_case_images(case_data, user_id)
        # Return the mutated caseData (images are set directly on the object)
        return case_data
    except Exception as e:
        print(f"[Gemini Route] image/pregenerate error: {e}")
        return JSONResponse({"error": str(e) or "Failed to pregenerate case images"}, status_code=500)


@router.post("/image/create")
async def image_create(request: Request):
    body = await request.json()
    try:
        result = await create_image_from_prompt(body.get("userPrompt", ""), body.get("aspectRatio"))
        return {"base64": result}
    except Exception as e:
        print(f"[Gemini Route] image/create error: {e}")
        return JSONResponse({"error": str(e) or "Failed to create image"}, status_code=500)


@router.post("/image/edit")
async def image_edit(request: Request):
    body = await request.json()
    try:
        result = await edit_image_with_prompt(
            body.get("baseImageBase64", ""), body.get("userPrompt", ""),
            body.get("aspectRatio"),
        )
        return {"base64": result}
    except Exception as e:
        print(f"[Gemini Route] image/edit error: {e}")
        return JSONResponse({"error": str(e) or "Failed to edit image"}, status_code=500)


# --- TTS ENDPOINT ---

@router.post("/tts")
async def tts(request: Request):
    body = await request.json()
    try:
        base64_audio = await generate_tts(
            body.get("text", ""), body.get("voiceName", ""),
            body.get("stylePrompt"),
        )
        if base64_audio is None:
            return JSONResponse(
                {"error": "TTS is unavailable (skipped or not configured). Check voice selection and API key."},
                status_code=503,
            )
        return {"audio": base64_audio}
    except Exception as e:
        print(f"[Gemini Route] tts error: {e}")
        return JSONResponse({"error": str(e) or "Failed to generate TTS"}, status_code=500)
